File: lib/jobs/resumeParsing/sensitiveData.ts
import type { JobsAiResumeTextExtractionResponse } from "@/lib/jobs/ai";

const REDACTED = "[redacted]";
const MAX_SAFE_MESSAGE_LENGTH = 500;

export interface SafeResumeExtractionMetadata {
  status: string;
  pageCount: number | null;
  detectedMimeType: string;
  characterCount: number;
  extractionMethod: string;
  warnings: string[];
  requiresOcr: boolean;
}

export function buildSafeResumeExtractionMetadata(
  extraction: JobsAiResumeTextExtractionResponse,
): SafeResumeExtractionMetadata {
  const { text: _discardedRawText, warnings, ...metadata } = extraction;

  return {
    ...metadata,
    warnings: warnings.map(sanitizeResumeProcessingMessage),
  };
}

export function sanitizeResumeProcessingMessage(message: string): string {
  return redactSensitiveResumeText(message)
    .replace(/\s+/g, " ")
    .trim()
    .slice(0, MAX_SAFE_MESSAGE_LENGTH);
}

export function redactSensitiveResumeText(value: string): string {
  return value
    .replace(/\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi, REDACTED)
    .replace(/\b(?:\+?\d[\s().-]?){8,}\d\b/g, REDACTED)
    .replace(/\bhttps?:\/\/\S+/gi, REDACTED)
    .replace(/\b(?:Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+/gi, REDACTED)
    .replace(/\bX-Amz-[A-Za-z0-9-]+=[^&\s]+/g, `X-Amz-${REDACTED}`)
    .replace(/\b[A-Za-z0-9+/]{80,}={0,2}\b/g, REDACTED);
}

File: services/jobs-ai/app/services/sensitive_data.py
import re

REDACTED = "[redacted]"


def sanitize_processing_message(message: str, max_length: int = 500) -> str:
    return re.sub(r"\s+", " ", redact_sensitive_text(message)).strip()[:max_length]


def redact_sensitive_text(value: str) -> str:
    redacted = re.sub(
        r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b",
        REDACTED,
        value,
        flags=re.IGNORECASE,
    )
    redacted = re.sub(r"\b(?:\+?\d[\s().-]?){8,}\d\b", REDACTED, redacted)
    redacted = re.sub(r"\bhttps?://\S+", REDACTED, redacted, flags=re.IGNORECASE)
    redacted = re.sub(
        r"\b(?:Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+",
        REDACTED,
        redacted,
        flags=re.IGNORECASE,
    )
    redacted = re.sub(r"\b[A-Za-z0-9+/]{80,}={0,2}\b", REDACTED, redacted)
    return redacted
